reflectapi/src/builder/handler.rs

```diff
@@ -278,10 +278,43 @@ where
             }
         };
 
+        let code = match output {
+            UntaggedResult::Ok(_) => http::StatusCode::OK,
+            UntaggedResult::Err(err) => {
+                let custom_error = err.status_code();
+                if custom_error == http::StatusCode::OK {
+                    // It means a user has implemented ToStatusCode trait for their
+                    // type incorrectly. It is a protocol error to return 200 status
+                    // code for an error response, as the client will not be able
+                    // to "cast" the response body to the correct type.
+                    // So, we are reverting it to internal error
+                    http::StatusCode::INTERNAL_SERVER_ERROR
+                } else {
+                    custom_error
+                }
+            }
+        };
+
         HandlerOutput {
-            code: output.status_code(),
+            code,
             body: bytes::Bytes::from(output_serialized),
             headers: response_headers,
         }
     }
 }
+
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize)]
+#[serde(untagged)]
+enum UntaggedResult<T, E> {
```

Comment:
Looking at this again, I really thing we do not need this extra type. It can be simply embedded into here (which is only 1 line down)
![image](https://github.com/user-attachments/assets/876e1ba2-7a4d-4ec7-827c-36f4456b1edf)

Reply:
If I'm not mistaken, I would have to copy paste the serialization logic into the `Ok` and `Err`branches? Could you give a code sample not sure I'm following.

Reply:
Yes, repeat to_vec_pretty call in both branches of the match statement. One will need to do it for Ok value, another for error value.

Reply:
I don't understand how that is a more appealing solution. This is efficient (compiler probably can just transmute if it's smart enough) and fairly clean.

Reply:
This is a matter of preference, not going to argue here. You can leave it as is if it is clearer. In my opinion explicit serialization for either result or error would be clearer to read for new developer unfamiliar with untagged story. As it that solution was not immediately obvious to you, it would not be obvious immediately clear to new reader. Although new wrapping struct approach is maybe more rusty way, which is also questionable.

Reply:
I'm objecting mainly because it's not just a few lines of duplication. It's the msgpack/json branch and the serialization failure handling branch too. I think I will leave as the extra type for now as you're willing to accept it.

Reply:
ok.. msgpack is not a strong reason, but I am ok with that, we need to move on with it :)

Reply:
The main thing I didn't want to copy is the following. Anyway, I will commit to the decision for now. Can revisit in the future.

```rust
        let output_serialized = match output_serialized {
            Ok(r) => r,
            Err(err) => {
                response_headers
                    .insert("content-type", http::HeaderValue::from_static("text/plain"));
                return HandlerOutput {
                    code: http::StatusCode::INTERNAL_SERVER_ERROR,
                    body: bytes::Bytes::from(
                        format!("Failed to serialize response body: {err}").into_bytes(),
                    ),
                    headers: response_headers,
                };
            }
        };
```